xds/src/test/java/io/grpc/xds/LookasideLbTest.java

```diff
@@ -111,7 +114,7 @@
 @RunWith(JUnit4.class)
 public class LookasideLbTest {
```

Comment:
The tests seem to a "lie", not even for the proof of interactions. The tests seem to make too much assumption on implementation such that all they are showing is passing an argument from one place to another, and that's it. I am not going to comment too much on this unit test, as I am not good at it and this test seems to be not easy (but I do believe some of the hardness comes from the inappropriateness of class structures, design/implementation decisions).

Btw, do you think passing in a pool for real `XdsClientImpl` instance in `ResolvedAddresses` to test the follow of CDS/EDS (without testing anything about load reporting) feasible and could make this better?

Reply:
What is proof of interactions in your mind? As this class depends on `LocalityStore`, we need the assumption on implementation _must_ pass arguments to LocalityStore and/or its factory. As long as `LocalityStore` is a dependency, no matter how implementation detail is changed, it should always pass exactly the same arguments to them. The arguments passed to `LocalityStore` is now the output of the balancer.

The real XdsNameResolver/CDS/EDS flow is heavy weight with dependencies, passing in a pool for real XdsClientImpl instance in ResolvedAddresses is not that real, you can only test a special XdsClientImpl instance with a trivial flow and it's still too heavy, sounds no better that testing a generic XdsClient instance.

Reply:
This test class is hardly readable and in a lot of places you are not verifying behaviors of ___code under test___, but just the test code itself (e.g., verify the side effects of mock/fake objects, that's the test code tens of lines above). I don't feel like this is right.

I am suggesting that you can create a real `XdsClientImpl` and pass it to `LookasideLb` via `ResolvedAddresses`. This unifies your tests for two separate use cases. 

Still, given the current structure, it will be tricky to test correctly. Previously we discussed to merge the implementation of `LocalityStore` into `LookasideLb`, that will make tests easier to be performed correctly (instead of verifying delegating method calls to a mocked object). For now, the ending outcome of `LookasideLb` would just be invoking some `LocalityStore` method, anything more than that is not part of code under test.


I think you don't need to worry about testing load reporting too much for now (we will have a pending change right?), focus on load balancing tests.